experimental/async_impl.hpp

```diff
@@ -0,0 +1,565 @@
+    {
+        __reduce_event = __exec.queue().submit([&](sycl::handler& __cgh) {
+            __cgh.depends_on(__reduce_event);
+
+            oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); // get an access to data under SYCL buffer
+            auto __temp_1_acc = __buf_1_ptr->template get_access<access_mode::read_write>(__cgh);
+            auto __temp_2_acc = __buf_2_ptr->template get_access<access_mode::write>(__cgh);
+            sycl::accessor<_Tp, 1, access_mode::read_write, sycl::access::target::local> __temp_local(
+                sycl::range<1>(__wgroup_size), __cgh);
+            __cgh.parallel_for<__kernel_name_t>(
+#if _ONEDPL_COMPILE_KERNEL
+                __kernel,
+#endif
+                sycl::nd_range<1>(sycl::range<1>(__n_groups * __wgroup_size), sycl::range<1>(__wgroup_size)),
+                [=](sycl::nd_item<1> __item_id) {
+                    auto __global_idx = __item_id.get_global_id(0);
+                    auto __local_idx = __item_id.get_local_id(0);
+                    // 1. Initialization (transform part). Fill local memory
+                    if (__is_first)
+                    {
+                        __u(__item_id, __global_idx, __n, __temp_local, __rngs...);
+                    }
+                    else
+                    {
+                        if (__global_idx < __n)
+                            __temp_local[__local_idx] = __temp_1_acc[__global_idx];
+                        __item_id.barrier(sycl::access::fence_space::local_space);
+                    }
+                    // 2. Reduce within work group using local memory
+                    auto __res = __brick_reduce(__item_id, __global_idx, __n, __temp_local);
+                    if (__local_idx == 0)
+                    {
+                        __temp_2_acc[__item_id.get_group(0)] = __res;
+                    }
+                });
+        });
+        if (__is_first)
+        {
+            __buf_2_ptr = &__temp_2;
+            __is_first = false;
+        }
+        else
+        {
+            ::std::swap(__buf_1_ptr, __buf_2_ptr);
+        }
+        __n = __n_groups;
+        __n_groups = (__n - 1) / __wgroup_size + 1;
+    } while (__n > 1);
+
+    // point of syncronization (on host access)
+    // return __buf_1_ptr->template get_access<access_mode::read>()[0];
+    return ::oneapi::dpl::__internal::__future<_Tp>( __reduce_event, *__buf_1_ptr, *__buf_2_ptr);
+}
+
+} // namespace __par_backend_hetero
+
+namespace __internal
+{
+
+// Async pattern overloads:
+
+// From: oneDPL/include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
+
+template <typename _ExecutionPolicy, typename _ForwardIterator, typename _Function>
+oneapi::dpl::__internal::__enable_if_async_execution_policy<_ExecutionPolicy,
+                                                            oneapi::dpl::__internal::__future<void>>
+__pattern_walk1_async(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last, _Function __f,
```

Comment:
These functions should be in `algorithm_impl_hetero.h`

Reply:
we agreed offline to keep it in separate files for now.